runtime/compiler/control/JITClientCompilationThread.cpp

```diff
@@ -499,7 +498,26 @@ handleServerMessage(JITServer::ClientStream *client, TR_J9VM *fe, JITServer::Mes
          vmInfo._floatInvokeExactThunkHelper = comp->getSymRefTab()->findOrCreateRuntimeHelper(TR_icallVMprJavaSendInvokeExactF, false, false, false)->getMethodAddress();
          vmInfo._doubleInvokeExactThunkHelper = comp->getSymRefTab()->findOrCreateRuntimeHelper(TR_icallVMprJavaSendInvokeExactD, false, false, false)->getMethodAddress();
          vmInfo._interpreterVTableOffset = TR::Compiler->vm.getInterpreterVTableOffset();
-         client->write(response, vmInfo);
+
+         // For multi-layered SCC support
+         vmInfo._j9SharedClassCacheDescriptorList = NULL;
+         std::vector<uintptr_t> listOfCacheStartAddress;
+         std::vector<uintptr_t> listOfCacheSizeBytes;
+         if (fe->sharedCache() && fe->sharedCache()->getCacheDescriptorList())
+            {
+            // The cache descriptor list is linked last to first and is circular, so last->previous == first.
+            J9SharedClassCacheDescriptor *firstCache = fe->sharedCache()->getCacheDescriptorList()->previous;
+            J9SharedClassCacheDescriptor *curCache = firstCache;
+            do
+               {
+               listOfCacheStartAddress.push_back((uintptr_t)curCache->cacheStartAddress);
+               listOfCacheSizeBytes.push_back(curCache->cacheSizeBytes);
+               curCache = curCache->previous;
+               }
+            while (curCache != firstCache);
+            }
```

Comment:
I haven't gone through the whole thing yet, but there are a couple of comments about this code.

1. The `previous` field only exists if `J9VM_OPT_MULTI_LAYER_SHARED_CLASS_CACHE` is defined, otherwise the cache descriptor list is a singly linked list with only a `next` field. You will have to guard any uses of the `previous` field with an `#ifdef` and have an `#else` part that does the right thing using `next`.
2. You don't actually need to walk the list backward for this operation. Just go forward, that way you don't need to use `previous` and you don't have to have an `#ifdef` and you can then also simplify `reconstructJ9SharedClassCacheDescriptorList()`.